client/CPlayerInterface.cpp

```diff
@@ -2905,3 +2914,52 @@ void CPlayerInterface::showWorldViewEx(const std::vector<ObjectPosInfo>& objectP
 
 	viewWorldMap();
 }
+
+void CPlayerInterface::updateAmbientSounds(bool resetAll)
+{
+	if(castleInt || battleInt || !makingTurn || !currentSelection)
+	{
+		CCS->soundh->ambientStopAllChannels();
+		return;
+	}
+	else if(!dynamic_cast<CAdvMapInt *>(GH.topInt()))
+	{
+		return;
+	}
+	if(resetAll)
+		CCS->soundh->ambientStopAllChannels();
+
+	std::map<std::string, int> currentSounds;
+	auto updateSounds = [&](std::string soundId, int distance) -> void
+	{
+		if(vstd::contains(currentSounds, soundId))
+			currentSounds[soundId] = std::max(currentSounds[soundId], distance);
+		else
+			currentSounds.insert(std::make_pair(soundId, distance));
+	};
+
+	int3 pos = currentSelection->getSightCenter();
+	std::unordered_set<int3, ShashInt3> tiles;
+	cb->getVisibleTilesInRange(tiles, pos, CCS->soundh->ambientGetRange(), int3::DIST_CHEBYSHEV);
+	for(int3 tile : tiles)
+	{
+		int dist = pos.dist(tile, int3::DIST_CHEBYSHEV);
+		// We want sound for every special terrain on tile and not just one on top
+		for(auto & ttObj : CGI->mh->ttiles[tile.x][tile.y][tile.z].objects)
+		{
+			auto obj = ttObj.obj;
+			if(!obj || !obj->getAmbientSound())
+				continue;

```

Comment:
This one I especially dislike and this check doesn't help. Problem is that map handler information might sill have object that already removed.

I not sure how exactly to avoid that.

Reply:
Copy sound-related data to `TerrainTileObject`.